### PR TITLE
feat: Return full `PhotoIdentifier` in save()

### PR DESCRIPTION
**Pull request by mrousavy**
# Summary

**TL;DR: Adds `saveAsset(): Promise<PhotoIdentifier>` to get the full photo asset immediately after saving.**

---

This PR changes the behaviour of the native `saveToCameraRoll` function to return the full `PhotoIdentifier` instead of just the URI.

The motivation behind this is feature request #568.

This is useful if you want to append the image to state and render it immediately, e.g. in Gallery or Camera apps (such as [ShadowLens](https://mrousavy.com/projects/shadowlens)), because you can use the `id` for equality comparisons and keys, and elements like width and height are immediately available.

To not make this breaking, `save()` will still return a string (the URI), and a new method called `saveAsset()` has been introduced, which returns the full photo asset.

## Test Plan

Save a photo by calling `save()` and `saveAsset()` and compare the different return values. one is a string, one the full PhotoIdentifier.

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ✅     |

## Checklist

- [x] I have tested this on a device and a simulator
- [x] I added the documentation in `README.md`
- [x] I updated the typed files (TS and Flow)
- [ ] I added a sample use of the API in the example project (`example/App.js`)

### CODE REVIEW

**Comment by Arjan-Zuidema** [comment]
@mrousavy I dont rely on the unit tests (because they did not work when I became a maintainer), I usually run tests on the FabricExample to make sure everything works.

**Comment by mrousavy** [comment]
Yep I think I'm done - I tested iOS and Android and both the old `save` and new `saveAsset` funcs work! :) 

I do think that the codebase is a bit tricky to work with, maybe we could tackle a Swift/Kotlin rewrite sometime in the next year and separate the components a bit cleaner. I could help with that

**Comment by santhoshvai** [comment]
That would be great

File: jest.setup.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ import {NativeModules} from 'react-native';
 
 // Mock the RNCNetInfo native module to allow us to unit test the JavaScript code
 NativeModules.CameraRollManager = {
-  saveToCameraRoll: jest.fn(),
+  saveToCameraRoll: jest.fn().mockResolvedValue({ node: { image: { uri: '' } } }),
```
Comment (idrisssakhi) [inline jest.setup.js:16]: i would say () => Promise.resolve(({ node: { image: { uri: '' } } })), so it can work

File: src/CameraRoll.ts [hunk 2]
```diff
@@ -229,13 +229,25 @@ export class CameraRoll {
   }
 
   /**
-   * Saves the photo or video to the camera roll or photo library.
+   * Saves the photo or video to the camera roll or photo library, and returns the URI of the newly created asset.
    *
    */
-  static save(
+  static async save(
```
Comment (idrisssakhi) [inline src/CameraRoll.ts:235]: i would put a deprecation comment here so we can remove it in the future

File: src/CameraRoll.ts [hunk 3]
```diff
@@ -229,13 +229,25 @@ export class CameraRoll {
...
   ): Promise<string> {
+    const asset = await this.saveAsset(tag, options);
+    return asset.node.image.uri;
+  }
+
+  /**
+   * Saves the photo or video to the camera roll or photo library, and returns the newly created asset.
+   *
+   */
+  static saveAsset(
```
Comment (idrisssakhi) [inline src/CameraRoll.ts:247]: please add the jsDoc params

File: src/__mocks__/nativeInterface.ts [hunk 4]
```diff
@@ -2,6 +2,6 @@
 
 module.exports = {
   deletePhotos: jest.fn(),
-  saveToCameraRoll: jest.fn(),
+  saveToCameraRoll: jest.fn().mockResolvedValue({node: {image: {uri: ''}}}),
```
Comment (idrisssakhi) [inline src/__mocks__/nativeInterface.ts:5]: save feedback as previous one for the mocking

File: src/__tests__/CameraRollTest.ts [hunk 5]
```diff
@@ -8,7 +8,9 @@ let mockGetPhotos: jest.Mock;
 
 jest.mock('../NativeCameraRollModule', () => {
   mockDeletePhotos = jest.fn();
-  mockSaveToCameraRoll = jest.fn();
+  mockSaveToCameraRoll = jest
+    .fn()
+    .mockResolvedValue({node: {image: {uri: ''}}});
```
Comment (idrisssakhi) [inline src/__tests__/CameraRollTest.ts:13]: same

**Review by idrisssakhi — CHANGES_REQUESTED** [review]

**Comment by idrisssakhi** [comment]
i think we should also update the PRs policy to only have one commit per PR eventhough we have the squash on github i find it better to have only one commit

**Comment by idrisssakhi** [comment]
@mrousavy seems ok after testing, if you take the feedbacks into account we can merge it after that

**Comment by mrousavy** [comment]
Done, applied your suggestions! :)

**Comment by mrousavy** [comment]
yuhuu, thanks! let me know if there's any issues/regressions coming up :)

**Comment by Arjan-Zuidema** [comment]
@idrisssakhi seems like the ci failed and no release was created
